
##### global workflow dependencies #####
# conda: "envs/global.yaml"

##### libraries #####
import os
import sys
import pandas as pd
import yaml
from snakemake.utils import min_version

##### set minimum snakemake version #####
min_version("8.25.3")

##### setup report #####
report: os.path.join("report", "workflow.rst")

##### load configs #####
configfile: os.path.join("config","config.yaml")

##### utility functions #####
# prepare configs for all workflows and datasets
config_wf = dict()
for ds in config["workflows"]:
    for wf in config["workflows"][ds]:
        with open(config["workflows"][ds][wf], 'r') as stream:
            try:
                config_wf[ds+'_'+wf]=yaml.safe_load(stream)
            except yaml.YAMLError as exc:
                print(exc)
    

##### load rules (one per dataset) #####
# include: os.path.join("rules", "quickstart.smk")
# include: os.path.join("rules", "digits.smk")
include: os.path.join("rules", "CorcesRNA.smk")
include: os.path.join("rules", "CorcesATAC.smk")
include: os.path.join("rules", "CorcesINT.smk")
include: os.path.join("rules", "Papalexi2021scCRISPR.smk")
include: os.path.join("rules", "docs.smk")
include: os.path.join("rules", "figures.smk")

#### Target Rule ####
rule all:
    input:
        # #### QUICKSTART ####
        # rules.quickstart_enrichment_analysis_all.input,
        # #### digits Analysis ####
        # rules.digits_unsupervised_analysis_all.input,
        #### RNA-seq Analysis Recipe (CorcesRNA) ####
        rules.CorcesRNA_fetch_ngs_all.input,
        rules.CorcesRNA_rnaseq_pipeline_all.input,
        rules.CorcesRNA_genome_tracks_all.input,
        rules.CorcesRNA_spilterlize_integrate_all.input,
        rules.CorcesRNA_unsupervised_analysis_all.input,
        rules.CorcesRNA_dea_limma_all.input,
        rules.CorcesRNA_enrichment_analysis_all.input,
        rules.CorcesRNA_reconstruct_lineage.output, # custom rule
        #### ATAC-seq Analysis Recipe (CorcesATAC) ####
        rules.CorcesATAC_fetch_ngs_all.input,
        rules.CorcesATAC_get_resources.output, # custom rule
        rules.CorcesATAC_atacseq_pipeline_all.input,
        rules.CorcesATAC_genome_tracks_all.input,
        rules.CorcesATAC_spilterlize_integrate_all.input,
        rules.CorcesATAC_unsupervised_analysis_all.input,
        rules.CorcesATAC_dea_limma_all.input,
        rules.CorcesATAC_enrichment_analysis_all.input,
        rules.CorcesATAC_reconstruct_lineage.output, # custom rule
        #### Integrative Analysis Recipe (CorcesINT) ####
        rules.CorcesINT_spilterlize_integrate_all.input,
        rules.CorcesINT_unsupervised_analysis_all.input,
        rules.CorcesINT_dea_limma_all.input,
        rules.CorcesINT_plot_correlation.output, # custom rule for divergent gene visualization
        rules.CorcesINT_enrichment_analysis_all.input,
        #### scCRISPR-seq Analysis Recipe (Papalexi2021scCRISPR) ####
        rules.Papalexi2021scCRISPR_get_data.output, # custom rule
        rules.Papalexi2021scCRISPR_scrnaseq_processing_seurat_all.input,
        rules.Papalexi2021scCRISPR_mixscape_seurat_all.input,
        rules.Papalexi2021scCRISPR_unsupervised_analysis_all.input,
        rules.Papalexi2021scCRISPR_dea_seurat_all.input,
        rules.Papalexi2021scCRISPR_enrichment_analysis_all.input,
        rules.Papalexi2021scCRISPR_functional_similarity.output, # custom rule
        #### Custom rules used only for docs/wiki (keep commented) ####
        # rules.CorcesRNA_plots.output,
        # rules.CorcesATAC_plots.output,
        # rules.CorcesINT_plots.output,
        # rules.Papalexi2021scCRISPR_plots.output,
        #### Figures ####
        rules.CorcesRNA_figures.output,
        rules.CorcesATAC_figures.output,
        rules.CorcesINT_figures.output,
        rules.Papalexi_figures.output,
    default_target: True # important for module usage
    threads: 1
    resources:
        mem_mb=8000,
    log:
        os.path.join("logs","rules","all.log")
